Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,11 +4,113 @@
 #include "Widgets/PSOverlayWidget.h"
 
 #include "Components/Image.h"
+#include "Curves/CurveFloat.h"
+#include "Data/PSDataAsset.h"
 
 #include UE_INLINE_GENERATED_CPP_BY_NAME(PSOverlayWidget)
 
-// Sets the visibility of the background overlay and lock icon.
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		ESlateVisibility PrevOverlayOpacity = PSCBackgroundOverlay->GetVisibility();
+		ESlateVisibility PrevIconOpacity = PSCBackgroundIconLock->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate && PrevIconOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+	if (bShouldPlayFadeAnimationInternal)
+	{
+		const bool bIsFinished = !FadeOverlayElementsAnimation(StartTimeFadeAnimationInternal);
+		if (bIsFinished)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+			StartTimeFadeAnimationInternal = 0.f;
+			if (!bIsFadeInAnimationInternal)
+			{
+				SetOverlayItemsVisibility(ESlateVisibility::Collapsed);
+			}
+		}
+	}
+}
+
+// Event to execute when widget is ready
+void UPSOverlayWidget::NativeConstruct()
+{
+	Super::NativeConstruct();
+	FadeCurveFloatInternal = UPSDataAsset::Get().GetFadeCurveFloat();
+}
+
+// Play the overlay elemetns fade-in/fade-out animation. Uses the internal FadeCurveFloatInternal initialized in NativeConstruct
+bool UPSOverlayWidget::FadeOverlayElementsAnimation(float& StartTimeRef)
+{
+	if (!StartTimeRef || !FadeCurveFloatInternal || !bShouldPlayFadeAnimationInternal)
+	{
+		StartTimeRef = 0.f;
+		return false;
+	}
+
+	const UWorld* World = GetWorld();
+	check(World);
+
+	const float SecondsSinceStart = GetWorld()->GetTimeSeconds() - StartTimeRef;
+	float OpacityValue = FadeCurveFloatInternal->GetFloatValue(SecondsSinceStart);
+
+	float MinTime = 0.f;
+	float MaxTime = 0.f;
+	FadeCurveFloatInternal->GetTimeRange(MinTime, MaxTime);
```

Comment:
Curbe might crash here if designer did not set. Please add ensuremsgf

Reply:
added: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/24cef2404f3f0b8519a6d6c8155c41f4e64530e0